IndexOutOfBoundsException on ForwardingCollection.java in guava.

IndexOutOfBoundsException on ForwardingCollection.java (https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/ForwardingCollection.java#L120)

src/main/java/com/google/common/collect/ForwardingCollection.java:120: error: An unhandled exception was thrown by the Error Prone static analysis plugin.
    return delegate().toArray();
                             ^
     Please report this at https://github.com/google/error-prone/issues/new and include the following:
  
     error-prone version: 2.1.1
     Stack Trace:
     java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
  	at com.sun.tools.javac.util.List.get(List.java:490)
  	at com.uber.nullaway.NullAway.handleInvocation(NullAway.java:987)
  	at com.uber.nullaway.NullAway.matchMethodInvocation(NullAway.java:297)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:907)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethodInvocation(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1644)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
  	at com.sun.source.util.TreeScanner.visitReturn(TreeScanner.java:469)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:1021)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitReturn(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1548)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
  	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  	at com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:530)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitBlock(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
  	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  	at com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:898)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitMethod(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
  	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  	at com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:590)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitClass(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:82)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:42)
  	at com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  	at com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  	at com.sun.source.util.TreeScanner.visitCompilationUnit(TreeScanner.java:144)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:605)
  	at com.google.errorprone.scanner.ErrorProneScanner.visitCompilationUnit(ErrorProneScanner.java:146)
  	at com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:591)
  	at com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  	at com.google.errorprone.scanner.Scanner.scan(Scanner.java:64)
  	at com.google.errorprone.scanner.ErrorProneScannerTransformer.apply(ErrorProneScannerTransformer.java:41)
  	at com.google.errorprone.ErrorProneAnalyzer.finished(ErrorProneAnalyzer.java:145)
  	at com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
  	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1425)
  	at com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1384)
  	at com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1523)
  	at com.sun.tools.javac.main.JavaCompiler.desugar(JavaCompiler.java:1439)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  	at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:100)
  	at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
  	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
  	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
  	at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:137)
  	at com.google.errorprone.BaseErrorProneCompiler.run(BaseErrorProneCompiler.java:108)
  	at com.google.errorprone.ErrorProneCompiler.run(ErrorProneCompiler.java:118)
  	at com.google.errorprone.ErrorProneCompiler.compile(ErrorProneCompiler.java:65)
  	at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne$CompilerInvoker.compile(JavacCompilerWithErrorProne.java:219)
  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.lang.reflect.Method.invoke(Method.java:498)
  	at org.codehaus.plexus.compiler.javac.errorprone.JavacCompilerWithErrorProne.performCompile(JavacCompilerWithErrorProne.java:91)
  	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:884)
  	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:129)
  	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
  	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
  	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
  	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
  	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
  	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
  	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
  	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
  	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.lang.reflect.Method.invoke(Method.java:498)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
  	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)

--- a/gradle.properties
+++ b/gradle.properties
@@ -13,7 +13,7 @@
 # org.gradle.parallel=true
 
 GROUP=com.uber.nullaway
-VERSION_NAME=0.2.3-SNAPSHOT
+VERSION_NAME=0.3.0-SNAPSHOT
 
 POM_DESCRIPTION=A fast annotation-based null checker for Java
 

--- a/nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
+++ b/nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
@@ -32,7 +32,7 @@ import com.sun.tools.javac.code.Types;
 import java.util.Objects;
 import javax.annotation.Nullable;
 
-/** Provides models for library routines for the null checker */
+/** Provides models for library routines for the null checker. */
 public interface LibraryModels {
 
   /**
@@ -61,7 +61,25 @@ public interface LibraryModels {
   /** @return set of library methods that may return null */
   ImmutableSet<MethodRef> nullableReturns();
 
-  /** representation of a method as a qualified class name + a signature for the method */
+  /**
+   * representation of a method as a qualified class name + a signature for the method
+   *
+   * <p>The formatting of a method signature should match the result of calling {@link
+   * Symbol.MethodSymbol#toString()} on the corresponding symbol. See {@link
+   * com.uber.nullaway.handlers.LibraryModelsHandler.DefaultLibraryModels} for examples. Basic
+   * principles:
+   *
+   * <ul>
+   *   <li>signature is a method name plus argument types, e.g., <code>foo(java.lang.Object,
+   *  java.lang.String)</code>
+   *   <li>constructor for class Foo looks like <code>Foo(java.lang.String)</code>
+   *   <li>If the method has its own type parameters, they need to be declared, like <code>
+   *       &lt;T&gt;checkNotNull(T)</code>
+   *   <li>Type bounds matter for generics, e.g., <code>addAll(java.lang.Iterable&lt;? extends
+   *   E&gt;)
+   *  </code>
+   * </ul>
+   */
   final class MethodRef {
 
     public final String clazz;
@@ -72,13 +90,23 @@ public interface LibraryModels {
       this.methodSig = methodSig;
     }
 
+    /**
+     * @param clazz containing class
+     * @param methodSig method signature in the appropriate format (see class docs)
+     * @return corresponding {@link MethodRef}
+     */
     public static MethodRef methodRef(Class<?> clazz, String methodSig) {
       return methodRef(clazz.getName(), methodSig);
     }
 
+    /**
+     * @param clazz containing class
+     * @param methodSig method signature in the appropriate format (see class docs)
+     * @return corresponding {@link MethodRef}
+     */
     public static MethodRef methodRef(String clazz, String methodSig) {
       Preconditions.checkArgument(
-          isValidMethodSig(methodSig), methodSig + " is not a valid " + "method signature");
+          isValidMethodSig(methodSig), methodSig + " is not a valid method signature");
       return new MethodRef(clazz, methodSig);
     }
 

--- a/nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
+++ b/nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
@@ -22,6 +22,7 @@
 
 package com.uber.nullaway;
 
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableCollection;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSetMultimap;
@@ -38,7 +39,7 @@ public interface LibraryModels {
    * @return map from the names of null-rejecting methods to the indexes of the arguments that
    *     aren't permitted to be null.
    */
-  ImmutableSetMultimap<MemberName, Integer> failIfNullParameters();
+  ImmutableSetMultimap<MethodRef, Integer> failIfNullParameters();
 
   /**
    * @return map from the names of methods with @NonNull parameters to the indexes of the arguments
@@ -49,57 +50,64 @@ public interface LibraryModels {
    *     #failIfNullParameters()}, it just learns that after the call the relevant parameters cannot
    *     be null.
    */
-  ImmutableSetMultimap<MemberName, Integer> nonNullParameters();
+  ImmutableSetMultimap<MethodRef, Integer> nonNullParameters();
 
   /**
    * @return map from the names of null-querying methods to the indexes of the arguments that are
    *     compared against null.
    */
-  ImmutableSetMultimap<MemberName, Integer> nullImpliesTrueParameters();
+  ImmutableSetMultimap<MethodRef, Integer> nullImpliesTrueParameters();
 
   /** @return set of library methods that may return null */
-  ImmutableSet<MemberName> nullableReturns();
+  ImmutableSet<MethodRef> nullableReturns();
 
-  /** representation of a member name as a qualified class name + a name for the member */
-  final class MemberName {
+  /** representation of a method as a qualified class name + a signature for the method */
+  final class MethodRef {
 
     public final String clazz;
-    public final String member;
+    public final String methodSig;
 
-    private MemberName(String clazz, String member) {
+    private MethodRef(String clazz, String methodSig) {
       this.clazz = clazz;
-      this.member = member;
+      this.methodSig = methodSig;
     }
 
-    public static MemberName member(Class<?> clazz, String member) {
-      return member(clazz.getName(), member);
+    public static MethodRef methodRef(Class<?> clazz, String methodSig) {
+      return methodRef(clazz.getName(), methodSig);
     }
 
-    public static MemberName member(String clazz, String member) {
-      return new MemberName(clazz, member);
+    public static MethodRef methodRef(String clazz, String methodSig) {
+      Preconditions.checkArgument(
+          isValidMethodSig(methodSig), methodSig + " is not a valid " + "method signature");
+      return new MethodRef(clazz, methodSig);
     }
 
-    public static MemberName fromSymbol(Symbol.MethodSymbol symbol) {
-      return member(symbol.owner.getQualifiedName().toString(), symbol.getSimpleName().toString());
+    private static boolean isValidMethodSig(String methodSig) {
+      // some basic checking to make sure it's not just a method name
+      return methodSig.contains("(") && methodSig.contains(")");
+    }
+
+    public static MethodRef fromSymbol(Symbol.MethodSymbol symbol) {
+      return methodRef(symbol.owner.getQualifiedName().toString(), symbol.toString());
     }
 
     @Override
     public boolean equals(Object obj) {
-      if (obj instanceof MemberName) {
-        MemberName other = (MemberName) obj;
-        return clazz.equals(other.clazz) && member.equals(other.member);
+      if (obj instanceof MethodRef) {
+        MethodRef other = (MethodRef) obj;
+        return clazz.equals(other.clazz) && methodSig.equals(other.methodSig);
       }
       return false;
     }
 
     @Override
     public int hashCode() {
-      return Objects.hash(clazz, member);
+      return Objects.hash(clazz, methodSig);
     }
 
     @Override
     public String toString() {
-      return "MemberName{" + "clazz='" + clazz + '\'' + ", member='" + member + '\'' + '}';
+      return "MethodRef{" + "clazz='" + clazz + '\'' + ", methodSig='" + methodSig + '\'' + '}';
     }
   }
 
@@ -116,12 +124,12 @@ public interface LibraryModels {
 
     @Nullable
     private static Symbol.MethodSymbol methodInSet(
-        Symbol.MethodSymbol symbol, Types types, ImmutableCollection<MemberName> memberNames) {
-      if (memberNames.contains(MemberName.fromSymbol(symbol))) {
+        Symbol.MethodSymbol symbol, Types types, ImmutableCollection<MethodRef> memberNames) {
+      if (memberNames.contains(MethodRef.fromSymbol(symbol))) {
         return symbol;
       }
       for (Symbol.MethodSymbol superSymbol : ASTHelpers.findSuperMethods(symbol, types)) {
-        if (memberNames.contains(MemberName.fromSymbol(superSymbol))) {
+        if (memberNames.contains(MethodRef.fromSymbol(superSymbol))) {
           return superSymbol;
         }
       }

--- a/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
+++ b/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
@@ -156,7 +156,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
 
     private static final ImmutableSetMultimap<MethodRef, Integer> FAIL_IF_NULL_PARAMETERS =
         new ImmutableSetMultimap.Builder<MethodRef, Integer>()
-            .put(MethodRef.methodRef(Preconditions.class, "<T>checkNotNull(T)"), 0)
+            .put(methodRef(Preconditions.class, "<T>checkNotNull(T)"), 0)
             .build();
 
     private static final ImmutableSetMultimap<MethodRef, Integer> NON_NULL_PARAMETERS =
@@ -164,14 +164,12 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
             .put(
                 methodRef(
                     "com.android.sdklib.build.ApkBuilder",
-                    "ApkBuilder(java.io.File,java"
-                        + ".io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
+                    "ApkBuilder(java.io.File,java.io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
                 0)
             .put(
                 methodRef(
                     "com.android.sdklib.build.ApkBuilder",
-                    "ApkBuilder(java.io.File,java"
-                        + ".io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
+                    "ApkBuilder(java.io.File,java.io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
                 1)
             .put(methodRef("com.google.common.collect.ImmutableList.Builder", "add(E)"), 0)
             .put(
@@ -183,23 +181,23 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
             .put(
                 methodRef(
                     "com.google.common.collect.ImmutableSet.Builder",
-                    "addAll(java.lang" + ".Iterable<? extends E>)"),
+                    "addAll(java.lang.Iterable<? extends E>)"),
                 0)
             .put(methodRef("com.google.common.collect.ImmutableSortedSet.Builder", "add(E)"), 0)
             .put(
                 methodRef(
                     "com.google.common.collect.ImmutableSortedSet.Builder",
-                    "addAll(java" + ".lang" + ".Iterable<? extends E>)"),
+                    "addAll(java.lang.Iterable<? extends E>)"),
                 0)
             .put(
                 methodRef(
                     "com.google.common.collect.Iterables",
-                    "<T>getFirst(java" + ".lang.Iterable<? extends T>,T)"),
+                    "<T>getFirst(java.lang.Iterable<? extends T>,T)"),
                 0)
             .put(
                 methodRef(
                     "com.google.common.util.concurrent.SettableFuture",
-                    "setException" + "(java.lang.Throwable)"),
+                    "setException(java.lang.Throwable)"),
                 0)
             .put(methodRef("java.io.File", "File(java.lang.String)"), 0)
             .put(methodRef("java.lang.Class", "getResource(java.lang.String)"), 0)
@@ -207,8 +205,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
             .put(methodRef("java.lang.System", "getProperty(java.lang.String)"), 0)
             .put(
                 methodRef(
-                    "java.net.URLClassLoader",
-                    "newInstance(java.net.URL[],java.lang" + ".ClassLoader)"),
+                    "java.net.URLClassLoader", "newInstance(java.net.URL[],java.lang.ClassLoader)"),
                 0)
             .put(
                 methodRef(
@@ -216,18 +213,16 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
                 0)
             .put(
                 methodRef(
-                    "javax.lang.model.util.Elements",
-                    "getPackageElement(java.lang" + ".CharSequence)"),
+                    "javax.lang.model.util.Elements", "getPackageElement(java.lang.CharSequence)"),
+                0)
+            .put(
+                methodRef(
+                    "javax.lang.model.util.Elements", "getTypeElement(java.lang.CharSequence)"),
                 0)
             .put(
                 methodRef(
                     "javax.lang.model.util.Elements",
-                    "getTypeElement(java.lang" + ".CharSequence)"),
-                0)
-            .put(
-                methodRef(
-                    "javax.lang.model.util.Elements",
-                    "getDocComment(javax.lang.model" + ".element.Element)"),
+                    "getDocComment(javax.lang.model.element.Element)"),
                 0)
             .put(methodRef("java.util.Deque", "addFirst(E)"), 0)
             .put(methodRef("java.util.Deque", "addLast(E)"), 0)
@@ -249,7 +244,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
 
     private static final ImmutableSetMultimap<MethodRef, Integer> NULL_IMPLIES_TRUE_PARAMETERS =
         new ImmutableSetMultimap.Builder<MethodRef, Integer>()
-            .put(MethodRef.methodRef(Strings.class, "isNullOrEmpty(java.lang.String)"), 0)
+            .put(methodRef(Strings.class, "isNullOrEmpty(java.lang.String)"), 0)
             .put(methodRef("android.text.TextUtils", "isEmpty(java.lang.String)"), 0)
             .build();
 

--- a/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
+++ b/nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
@@ -22,7 +22,7 @@
 
 package com.uber.nullaway.handlers;
 
-import static com.uber.nullaway.LibraryModels.MemberName.member;
+import static com.uber.nullaway.LibraryModels.MethodRef.methodRef;
 import static com.uber.nullaway.Nullness.NONNULL;
 
 import com.google.common.base.Preconditions;
@@ -74,9 +74,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
       ImmutableSet<Integer> nonNullPositions) {
     return Sets.union(
             nonNullPositions,
-            libraryModels
-                .nonNullParameters()
-                .get(LibraryModels.MemberName.fromSymbol(methodSymbol)))
+            libraryModels.nonNullParameters().get(LibraryModels.MethodRef.fromSymbol(methodSymbol)))
         .immutableCopy();
   }
 
@@ -113,7 +111,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
       List<Node> arguments,
       Symbol.MethodSymbol callee) {
     Set<Integer> nullImpliesTrueParameters =
-        libraryModels.nullImpliesTrueParameters().get(LibraryModels.MemberName.fromSymbol(callee));
+        libraryModels.nullImpliesTrueParameters().get(LibraryModels.MethodRef.fromSymbol(callee));
     for (AccessPath accessPath : accessPathsAtIndexes(nullImpliesTrueParameters, arguments)) {
       elseUpdates.set(accessPath, NONNULL);
     }
@@ -140,7 +138,7 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
       List<Node> arguments,
       Symbol.MethodSymbol callee) {
     Set<Integer> requiredNonNullParameters =
-        libraryModels.failIfNullParameters().get(LibraryModels.MemberName.fromSymbol(callee));
+        libraryModels.failIfNullParameters().get(LibraryModels.MethodRef.fromSymbol(callee));
     for (AccessPath accessPath : accessPathsAtIndexes(requiredNonNullParameters, arguments)) {
       bothUpdates.set(accessPath, NONNULL);
     }
@@ -156,125 +154,174 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
 
   private static class DefaultLibraryModels implements LibraryModels {
 
-    private static final ImmutableSetMultimap<MemberName, Integer> FAIL_IF_NULL_PARAMETERS =
-        new ImmutableSetMultimap.Builder<MemberName, Integer>()
-            .put(member(Preconditions.class, "checkNotNull"), 0)
+    private static final ImmutableSetMultimap<MethodRef, Integer> FAIL_IF_NULL_PARAMETERS =
+        new ImmutableSetMultimap.Builder<MethodRef, Integer>()
+            .put(MethodRef.methodRef(Preconditions.class, "<T>checkNotNull(T)"), 0)
             .build();
 
-    private static final ImmutableSetMultimap<MemberName, Integer> NON_NULL_PARAMETERS =
-        new ImmutableSetMultimap.Builder<MemberName, Integer>()
-            .put(member("com.android.sdklib.build.ApkBuilder", "<init>"), 0)
-            .put(member("com.android.sdklib.build.ApkBuilder", "<init>"), 1)
-            .put(member("com.android.util.CommandLineParser$Mode", "process"), 0)
-            .put(member("com.google.common.base.Objects$ToStringHelper", "add"), 0)
-            .put(member("com.google.common.collect.ImmutableList$Builder", "add"), 0)
-            .put(member("com.google.common.collect.ImmutableList$Builder", "addAll"), 0)
-            .put(member("com.google.common.collect.ImmutableSortedSet$Builder", "add"), 0)
-            .put(member("com.google.common.collect.Iterables", "getFirst"), 0)
-            .put(member("com.google.common.util.concurrent.SettableFuture", "setException"), 0)
-            .put(member("java.io.File", "<init>"), 0)
-            .put(member("java.lang.Class", "getResource"), 0)
-            .put(member("java.lang.Class", "isAssignableFrom"), 0)
-            .put(member("java.lang.System", "getProperty"), 0)
-            .put(member("java.net.URLClassLoader", "newInstance"), 0)
-            .put(member("javax.lang.model.element.Element", "getAnnotation"), 0)
-            .put(member("javax.lang.model.util.Elements", "getPackageElement"), 0)
-            .put(member("javax.lang.model.util.Elements", "getTypeElement"), 0)
-            .put(member("javax.lang.model.util.Elements", "getDocComment"), 0)
-            .put(member("java.util.Deque", "addFirst"), 0)
-            .put(member("java.util.Deque", "addLast"), 0)
-            .put(member("java.util.Deque", "offerFirst"), 0)
-            .put(member("java.util.Deque", "offerLast"), 0)
-            .put(member("java.util.Deque", "add"), 0)
-            .put(member("java.util.Deque", "offer"), 0)
-            .put(member("java.util.Deque", "push"), 0)
-            .put(member("java.util.Collection", "toArray"), 0)
-            .put(member("java.util.ArrayDeque", "addFirst"), 0)
-            .put(member("java.util.ArrayDeque", "addLast"), 0)
-            .put(member("java.util.ArrayDeque", "offerFirst"), 0)
-            .put(member("java.util.ArrayDeque", "offerLast"), 0)
-            .put(member("java.util.ArrayDeque", "add"), 0)
-            .put(member("java.util.ArrayDeque", "offer"), 0)
-            .put(member("java.util.ArrayDeque", "push"), 0)
-            .put(member("java.util.ArrayDeque", "toArray"), 0)
+    private static final ImmutableSetMultimap<MethodRef, Integer> NON_NULL_PARAMETERS =
+        new ImmutableSetMultimap.Builder<MethodRef, Integer>()
+            .put(
+                methodRef(
+                    "com.android.sdklib.build.ApkBuilder",
+                    "ApkBuilder(java.io.File,java"
+                        + ".io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
+                0)
+            .put(
+                methodRef(
+                    "com.android.sdklib.build.ApkBuilder",
+                    "ApkBuilder(java.io.File,java"
+                        + ".io.File,java.io.File,java.lang.String,java.io.PrintStream))"),
+                1)
+            .put(methodRef("com.google.common.collect.ImmutableList.Builder", "add(E)"), 0)
+            .put(
+                methodRef(
+                    "com.google.common.collect.ImmutableList.Builder",
+                    "addAll(java.lang.Iterable<? extends E>)"),
+                0)
+            .put(methodRef("com.google.common.collect.ImmutableSet.Builder", "add(E)"), 0)
+            .put(
+                methodRef(
+                    "com.google.common.collect.ImmutableSet.Builder",
+                    "addAll(java.lang" + ".Iterable<? extends E>)"),
+                0)
+            .put(methodRef("com.google.common.collect.ImmutableSortedSet.Builder", "add(E)"), 0)
+            .put(
+                methodRef(
+                    "com.google.common.collect.ImmutableSortedSet.Builder",
+                    "addAll(java" + ".lang" + ".Iterable<? extends E>)"),
+                0)
+            .put(
+                methodRef(
+                    "com.google.common.collect.Iterables",
+                    "<T>getFirst(java" + ".lang.Iterable<? extends T>,T)"),
+                0)
+            .put(
+                methodRef(
+                    "com.google.common.util.concurrent.SettableFuture",
+                    "setException" + "(java.lang.Throwable)"),
+                0)
+            .put(methodRef("java.io.File", "File(java.lang.String)"), 0)
+            .put(methodRef("java.lang.Class", "getResource(java.lang.String)"), 0)
+            .put(methodRef("java.lang.Class", "isAssignableFrom(java.lang.Class<?>)"), 0)
+            .put(methodRef("java.lang.System", "getProperty(java.lang.String)"), 0)
+            .put(
+                methodRef(
+                    "java.net.URLClassLoader",
+                    "newInstance(java.net.URL[],java.lang" + ".ClassLoader)"),
+                0)
+            .put(
+                methodRef(
+                    "javax.lang.model.element.Element", "<A>getAnnotation(java.lang.Class<A>)"),
+                0)
+            .put(
+                methodRef(
+                    "javax.lang.model.util.Elements",
+                    "getPackageElement(java.lang" + ".CharSequence)"),
+                0)
+            .put(
+                methodRef(
+                    "javax.lang.model.util.Elements",
+                    "getTypeElement(java.lang" + ".CharSequence)"),
+                0)
+            .put(
+                methodRef(
+                    "javax.lang.model.util.Elements",
+                    "getDocComment(javax.lang.model" + ".element.Element)"),
+                0)
+            .put(methodRef("java.util.Deque", "addFirst(E)"), 0)
+            .put(methodRef("java.util.Deque", "addLast(E)"), 0)
+            .put(methodRef("java.util.Deque", "offerFirst(E)"), 0)
+            .put(methodRef("java.util.Deque", "offerLast(E)"), 0)
+            .put(methodRef("java.util.Deque", "add(E)"), 0)
+            .put(methodRef("java.util.Deque", "offer(E)"), 0)
+            .put(methodRef("java.util.Deque", "push(E)"), 0)
+            .put(methodRef("java.util.Collection", "<T>toArray(T[])"), 0)
+            .put(methodRef("java.util.ArrayDeque", "addFirst(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "addLast(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "offerFirst(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "offerLast(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "add(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "offer(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "push(E)"), 0)
+            .put(methodRef("java.util.ArrayDeque", "<T>toArray(T[])"), 0)
             .build();
 
-    private static final ImmutableSetMultimap<MemberName, Integer> NULL_IMPLIES_TRUE_PARAMETERS =
-        new ImmutableSetMultimap.Builder<MemberName, Integer>()
-            .put(member(Strings.class, "isNullOrEmpty"), 0)
-            .put(member("android.text.TextUtils", "isEmpty"), 0)
+    private static final ImmutableSetMultimap<MethodRef, Integer> NULL_IMPLIES_TRUE_PARAMETERS =
+        new ImmutableSetMultimap.Builder<MethodRef, Integer>()
+            .put(MethodRef.methodRef(Strings.class, "isNullOrEmpty(java.lang.String)"), 0)
+            .put(methodRef("android.text.TextUtils", "isEmpty(java.lang.String)"), 0)
             .build();
 
-    private static final ImmutableSet<MemberName> NULLABLE_RETURNS =
-        new ImmutableSet.Builder<MemberName>()
-            .add(member("java.lang.ref.Reference", "get"))
-            .add(member("java.lang.ref.PhantomReference", "get"))
-            .add(member("java.lang.ref.SoftReference", "get"))
-            .add(member("java.lang.ref.WeakReference", "get"))
-            .add(member("java.util.concurrent.atomic.AtomicReference", "get"))
-            .add(member("java.util.Map", "get"))
-            .add(member("javax.lang.model.element.Element", "getEnclosingElement"))
-            .add(member("javax.lang.model.element.ExecutableElement", "getDefaultValue"))
-            .add(member("javax.lang.model.element.PackageElement", "getEnclosingElement"))
-            .add(member("javax.lang.model.element.VariableElement", "getConstantValue"))
-            .add(member("javax.lang.model.type.WildcardType", "getSuperBound"))
-            .add(member("android.app.ActivityManager", "getRunningAppProcesses"))
-            .add(member("android.view.View", "getHandler"))
-            .add(member("java.lang.Throwable", "getMessage"))
+    private static final ImmutableSet<MethodRef> NULLABLE_RETURNS =
+        new ImmutableSet.Builder<MethodRef>()
+            .add(methodRef("java.lang.ref.Reference", "get()"))
+            .add(methodRef("java.lang.ref.PhantomReference", "get()"))
+            .add(methodRef("java.lang.ref.SoftReference", "get()"))
+            .add(methodRef("java.lang.ref.WeakReference", "get()"))
+            .add(methodRef("java.util.concurrent.atomic.AtomicReference", "get()"))
+            .add(methodRef("java.util.Map", "get(java.lang.Object)"))
+            .add(methodRef("javax.lang.model.element.Element", "getEnclosingElement()"))
+            .add(methodRef("javax.lang.model.element.ExecutableElement", "getDefaultValue()"))
+            .add(methodRef("javax.lang.model.element.PackageElement", "getEnclosingElement()"))
+            .add(methodRef("javax.lang.model.element.VariableElement", "getConstantValue()"))
+            .add(methodRef("javax.lang.model.type.WildcardType", "getSuperBound()"))
+            .add(methodRef("android.app.ActivityManager", "getRunningAppProcesses()"))
+            .add(methodRef("android.view.View", "getHandler()"))
+            .add(methodRef("java.lang.Throwable", "getMessage()"))
             .build();
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> failIfNullParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> failIfNullParameters() {
       return FAIL_IF_NULL_PARAMETERS;
     }
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> nonNullParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> nonNullParameters() {
       return NON_NULL_PARAMETERS;
     }
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> nullImpliesTrueParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> nullImpliesTrueParameters() {
       return NULL_IMPLIES_TRUE_PARAMETERS;
     }
 
     @Override
-    public ImmutableSet<MemberName> nullableReturns() {
+    public ImmutableSet<MethodRef> nullableReturns() {
       return NULLABLE_RETURNS;
     }
   }
 
   private static class CombinedLibraryModels implements LibraryModels {
 
-    private final ImmutableSetMultimap<MemberName, Integer> fail_if_null_parameters;
+    private final ImmutableSetMultimap<MethodRef, Integer> fail_if_null_parameters;
 
-    private final ImmutableSetMultimap<MemberName, Integer> non_null_parameters;
+    private final ImmutableSetMultimap<MethodRef, Integer> non_null_parameters;
 
-    private final ImmutableSetMultimap<MemberName, Integer> null_implies_true_parameters;
+    private final ImmutableSetMultimap<MethodRef, Integer> null_implies_true_parameters;
 
-    private final ImmutableSet<MemberName> nullable_returns;
+    private final ImmutableSet<MethodRef> nullable_returns;
 
     public CombinedLibraryModels(Iterable<LibraryModels> models) {
-      ImmutableSetMultimap.Builder<MemberName, Integer> failIfNullParametersBuilder =
+      ImmutableSetMultimap.Builder<MethodRef, Integer> failIfNullParametersBuilder =
           new ImmutableSetMultimap.Builder<>();
-      ImmutableSetMultimap.Builder<MemberName, Integer> nonNullParametersBuilder =
+      ImmutableSetMultimap.Builder<MethodRef, Integer> nonNullParametersBuilder =
           new ImmutableSetMultimap.Builder<>();
-      ImmutableSetMultimap.Builder<MemberName, Integer> nullImpliesTrueParametersBuilder =
+      ImmutableSetMultimap.Builder<MethodRef, Integer> nullImpliesTrueParametersBuilder =
           new ImmutableSetMultimap.Builder<>();
-      ImmutableSet.Builder<MemberName> nullableReturnsBuilder = new ImmutableSet.Builder<>();
+      ImmutableSet.Builder<MethodRef> nullableReturnsBuilder = new ImmutableSet.Builder<>();
       for (LibraryModels libraryModels : models) {
-        for (Map.Entry<MemberName, Integer> entry :
-            libraryModels.failIfNullParameters().entries()) {
+        for (Map.Entry<MethodRef, Integer> entry : libraryModels.failIfNullParameters().entries()) {
           failIfNullParametersBuilder.put(entry);
         }
-        for (Map.Entry<MemberName, Integer> entry : libraryModels.nonNullParameters().entries()) {
+        for (Map.Entry<MethodRef, Integer> entry : libraryModels.nonNullParameters().entries()) {
           nonNullParametersBuilder.put(entry);
         }
-        for (Map.Entry<MemberName, Integer> entry :
+        for (Map.Entry<MethodRef, Integer> entry :
             libraryModels.nullImpliesTrueParameters().entries()) {
           nullImpliesTrueParametersBuilder.put(entry);
         }
-        for (MemberName name : libraryModels.nullableReturns()) {
+        for (MethodRef name : libraryModels.nullableReturns()) {
           nullableReturnsBuilder.add(name);
         }
       }
@@ -285,22 +332,22 @@ public class LibraryModelsHandler extends BaseNoOpHandler {
     }
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> failIfNullParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> failIfNullParameters() {
       return fail_if_null_parameters;
     }
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> nonNullParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> nonNullParameters() {
       return non_null_parameters;
     }
 
     @Override
-    public ImmutableSetMultimap<MemberName, Integer> nullImpliesTrueParameters() {
+    public ImmutableSetMultimap<MethodRef, Integer> nullImpliesTrueParameters() {
       return null_implies_true_parameters;
     }
 
     @Override
-    public ImmutableSet<MemberName> nullableReturns() {
+    public ImmutableSet<MethodRef> nullableReturns() {
       return nullable_returns;
     }
   }

--- a/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNativeModels.java
+++ b/nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayNativeModels.java
@@ -22,15 +22,23 @@
 
 package com.uber.nullaway.testdata;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
 import java.io.File;
 import java.lang.ref.WeakReference;
+import java.net.URLClassLoader;
 import java.util.ArrayDeque;
+import java.util.Collection;
 import java.util.Deque;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicReference;
+import javax.lang.model.element.Element;
+import javax.lang.model.util.Elements;
 
 public class NullAwayNativeModels {
 
@@ -194,9 +202,24 @@ public class NullAwayNativeModels {
   static void nonNullParameters() {
     // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
     NullAwayNativeModels.class.getResource(null);
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    NullAwayNativeModels.class.isAssignableFrom(null);
     String s = null;
     // BUG: Diagnostic contains: passing @Nullable parameter 's' where @NonNull is required
     File f = new File(s);
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    URLClassLoader.newInstance(null, NullAwayNativeModels.class.getClassLoader());
+  }
+
+  static void elementStuff(Element e, Elements elems) {
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    e.getAnnotation(null);
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    elems.getPackageElement(null);
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    elems.getTypeElement(null);
+    // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
+    elems.getDocComment(null);
   }
 
   static void arrayDequeStuff() {
@@ -217,6 +240,8 @@ public class NullAwayNativeModels {
     d.push(null);
     // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
     d.toArray(null);
+    // this should be fine
+    d.toArray();
   }
 
   static void dequeStuff() {
@@ -237,5 +262,24 @@ public class NullAwayNativeModels {
     d.push(null);
     // BUG: Diagnostic contains: passing @Nullable parameter 'null' where @NonNull is required
     d.toArray(null);
+  }
+
+  static void guavaStuff() {
+    Collection<String> c = null;
+    Object o = null;
+    // BUG: Diagnostic contains: passing @Nullable parameter 'c' where @NonNull is required
+    ImmutableList.builder().addAll(c).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'o' where @NonNull is required
+    ImmutableList.builder().add(o).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'c' where @NonNull is required
+    ImmutableSet.builder().addAll(c).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'o' where @NonNull is required
+    ImmutableSet.builder().add(o).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'c' where @NonNull is required
+    ImmutableSortedSet.builder().addAll(c).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'o' where @NonNull is required
+    ImmutableSortedSet.builder().add(o).build();
+    // BUG: Diagnostic contains: passing @Nullable parameter 'c' where @NonNull is required
+    Iterables.getFirst(c, "hi");
   }
 }

--- a/sample-library-model/src/main/java/com/uber/modelexample/ExampleLibraryModels.java
+++ b/sample-library-model/src/main/java/com/uber/modelexample/ExampleLibraryModels.java
@@ -15,7 +15,7 @@
  */
 package com.uber.modelexample;
 
-import static com.uber.nullaway.LibraryModels.MemberName.member;
+import static com.uber.nullaway.LibraryModels.MethodRef.methodRef;
 
 import com.google.auto.service.AutoService;
 import com.google.common.collect.ImmutableSet;
@@ -26,24 +26,24 @@ import com.uber.nullaway.LibraryModels;
 public class ExampleLibraryModels implements LibraryModels {
 
   @Override
-  public ImmutableSetMultimap<MemberName, Integer> failIfNullParameters() {
+  public ImmutableSetMultimap<MethodRef, Integer> failIfNullParameters() {
     return ImmutableSetMultimap.of();
   }
 
   @Override
-  public ImmutableSetMultimap<MemberName, Integer> nonNullParameters() {
+  public ImmutableSetMultimap<MethodRef, Integer> nonNullParameters() {
     return ImmutableSetMultimap.of();
   }
 
   @Override
-  public ImmutableSetMultimap<MemberName, Integer> nullImpliesTrueParameters() {
-    return ImmutableSetMultimap.<MemberName, Integer>builder()
-        .put(member("org.utilities.StringUtils", "isEmptyOrNull"), 0)
+  public ImmutableSetMultimap<MethodRef, Integer> nullImpliesTrueParameters() {
+    return ImmutableSetMultimap.<MethodRef, Integer>builder()
+        .put(methodRef("org.utilities.StringUtils", "isEmptyOrNull(java.lang.CharSequence)"), 0)
         .build();
   }
 
   @Override
-  public ImmutableSet<MemberName> nullableReturns() {
+  public ImmutableSet<MethodRef> nullableReturns() {
     return ImmutableSet.of();
   }
 }

